markdown/events/2021/bytesize-5-dsl2-module-development.md

```diff
@@ -32,3 +32,257 @@ This week, Harshil Patel ([@drpatelh](http://github.com/drpatelh/)) will present
+
+[21:35](https://www.youtube.com/watch?v=ggGGhTMgyHI&t=1295)
+By default, they're initialized to empty values like strings or false, but they can be overwritten and that's the key thing here. You can use those options to then overwrite where you're saving files, for example. You can also provide any non-mandatory arguments as strings to this module. And I think that's quite important in terms of flexibility, so all mandatory arguments (what we would consider mandatory arguments) that should be defined in the module are anything that that involves using inputs and output files because those (i sort of the ethos of Nextflow) need to be staged properly and put in the right place and they need to be defined as inputs and outputs.
+
+[22:14](https://www.youtube.com/watch?v=ggGGhTMgyHI&t=1334)
+Also, for example, anything where you can use Nextflow resource variables and define them in the script, like threads and so on. So, to show you how these are how these options are passed around (I mean, it took me a while to figure out how best to do this) but every pipeline will have a conf/modules.config and this is specifically within the pipeline repository and this will contain a list of modules that you have, along with arguments and and custom publishing options that that you that you may require.
+
+And in this case, I've just used a simple example where I've got fastqc, I've set the arguments to quiet and this is a non-mandatory argument, it's just a string that can be passed to the module.
+
+[22:56](https://www.youtube.com/watch?v=ggGGhTMgyHI&t=1376)
+I said publish it here and I've said, also in terms of the files that i want to publish, publish anything with a html dot extension in the top-level directory of my_fastqc and anything with a zip extension, publish it in a zip directory.
+
+And so, it's quite simplistic in the way that it works. Now this modules.config is then typically loaded in your Nextflow config and then all of these parameters then become available to the main script here as a result of this loading. But the great thing about this is now also that because this is a Nextflow params and this modules is just a groovy map, users can overwrite these modules if they want via their own custom configs. I think this is one of the key features here and one thing that I wanted to implement to make things as flexible as possible because with dsl1 pipelines you typically have to physically add a parameter to the pipeline if you want to amend the command line argument, for example.
```

Comment:
```suggestion
And so, it's quite simplistic in the way that it works. Now this modules.config is then typically loaded in your Nextflow config and then all of these parameters then become available to the main script here as a result of this loading. But the great thing about this is now also that because this is a Nextflow params and this modules is just a groovy map, users can overwrite these modules if they want via their own custom configs. I think this is one of the key features here and one thing that I wanted to implement to make things as flexible as possible because with DSL1 pipelines you typically have to physically add a parameter to the pipeline if you want to amend the command line argument, for example.
```